Templates/sparsh_table.cpp

```diff
@@ -0,0 +1,44 @@
+#include<bits/stdc++.h>
+using namespace std;
+/*
+Range sum query
+Provided with an integer array Arr of length N, answer Q queries:
+given L and R, 0 ≤ L, R < N, find ArrL + ArrL + 1 + … + ArrR.
+N and Q are up to 105, |Arri| ≤ 109.
+*/
+const int k = 16;
+const int N = 1e5;
+const int ZERO = 0; // ZERO + x = x + ZERO = x (for any x)
+
+long long table[N][k + 1]; // k + 1 because we need to access table[r][k]
+int Arr[N];
+
+int main()
+{
+    int n, L, R, q;
+    cin >> n; // array size
+    for(int i = 0; i < n; i++)
+        cin >> Arr[i];
+
+    // build Sparse Table
+    for(int i = 0; i < n; i++)
+        table[i][0] = Arr[i];
+    for(int j = 1; j <= k; j++) {
+        for(int i = 0; i <= n - (1 << j); i++)
+            table[i][j] = table[i][j - 1] + table[i + (1 << (j - 1))][j - 1];
```

Comment:
@bag-PacKer Can you add comments to explain the intuition/logic behind this line of code?